Firestore/core/src/firebase/firestore/local/leveldb_key.cc

```diff
@@ -108,301 +108,319 @@ enum ComponentLabel {
-      *contents = tmp;
-      *result = static_cast<int32_t>(raw_result);
-      return true;
+  void ReadTableNameMatching(const char *expected_table_name) {
+    if (!ReadLabeledStringMatching(ComponentLabel::TableName, expected_table_name)) {
+      Fail();
     }
   }
-  return false;
-}
 
-/**
- * Writes a component label and a signed integer to the given key destination.
- */
-void WriteLabeledInt32(std::string *dest, ComponentLabel label, int32_t value) {
-  WriteComponentLabel(dest, label);
-  OrderedCode::WriteSignedNumIncreasing(dest, value);
-}
+  model::BatchId ReadBatchId() {
+    return ReadLabeledInt32(ComponentLabel::BatchId);
+  }
 
-/**
- * Reads a component label and signed number from the given key contents and
- * verifies that the label matches the expected_label and the value fits in a
- * 32-bit integer.
- *
- * If the read is unsuccessful, the label didn't match, or the number that was
- * read was out of bounds for an int32_t, returns false, and changes none of its
- * arguments.
- *
- * If the read is successful, returns true, contents will be updated to the next
- * unread byte, and value will be set to the decoded integer value.
- */
-bool ReadLabeledInt32(leveldb::Slice *contents,
-                      ComponentLabel expected_label,
-                      int32_t *value) {
-  leveldb::Slice tmp = *contents;
-  if (ReadComponentLabelMatching(&tmp, expected_label)) {
-    if (ReadInt32(&tmp, value)) {
-      *contents = tmp;
-      return true;
-    }
+  std::string ReadCanonicalId() {
+    return ReadLabeledString(ComponentLabel::CanonicalId);
   }
-  return false;
-}
 
-/**
- * Writes a component label and an encoded string to the given key destination.
- */
-void WriteLabeledString(std::string *dest,
-                        ComponentLabel label,
-                        absl::string_view value) {
-  WriteComponentLabel(dest, label);
-  OrderedCode::WriteString(dest, value);
-}
+  model::TargetId ReadTargetId() {
+    return ReadLabeledInt32(ComponentLabel::TargetId);
+  }
 
-/**
- * Reads a component label and a string from the given key contents and verifies
- * that the label matches the expected_label.
- *
- * If the read is unsuccessful or the label didn't match, returns false, and
- * changes none of its arguments.
- *
- * If the read is successful, returns true, contents will be updated to the next
- * unread byte, and value will be set to the decoded string value.
- */
-bool ReadLabeledString(leveldb::Slice *contents,
-                       ComponentLabel expected_label,
-                       std::string *value) {
-  leveldb::Slice tmp = *contents;
-  if (ReadComponentLabelMatching(&tmp, expected_label)) {
-    if (ReadString(&tmp, value)) {
-      *contents = tmp;
-      return true;
+  std::string ReadUserId() {
+    return ReadLabeledString(ComponentLabel::UserId);
+  }
+
+  /** OrderedCode::ReadSignedNumIncreasing adapted to leveldb::Slice. */
+  int64_t ReadSignedNumIncreasing() {
```

Comment:
Do you think some of these lower-level functions could now be made `private`? I think it's also a small improvement that using a class can bring: previously, there also was a separation between these and domain-level functions (e.g., `ReadBatchId`), but it was only in the comments. Candidates for `private` are:
* `ReadSignedNumIncreasing`;
* `ReadString`;
* `ReadInt32`;
* (maybe) `ReadLabeledInt32`;
* (maybe) `WriteLabeledInt32`.

Reply:
I think nearly all the lowest-level ones could be made private if `Describe` were a member on the Reader. I'll attempt this next.